core/objects.go

```diff
@@ -358,17 +313,24 @@ func (ch Challenge) IsSane(completed bool) bool {
 // MergeResponse copies a subset of client-provided data to the current Challenge.
 // Note: This method does not update the challenge on the left side of the '.'
 func (ch Challenge) MergeResponse(resp Challenge) Challenge {
-	// Only override fields that are supposed to be client-provided
-	if len(ch.Path) == 0 {
-		ch.Path = resp.Path
-	}
-
-	if len(ch.S) == 0 {
-		ch.S = resp.S
-	}
+	switch ch.Type {
+	case ChallengeTypeSimpleHTTP:
+		// For simpleHttp, only "tls" is client-provided
+		// If "tls" is not provided, default to "true"
+		if resp.TLS != nil {
+			ch.TLS = resp.TLS
+		} else {
+			ch.TLS = new(bool)
+			*ch.TLS = true
```

Comment:
Just a simplification.  Treat the different types of challenges independently instead of requiring the reader to tease out what applies to what.